lib/otf/fontsense.cpp

```diff
@@ -0,0 +1,159 @@
+#include "fontsense.h"
+#include "unimath/uni_font.h"
+
+#include <stdlib.h>
+#include <queue>
+#include <string.h>
+#include <sstream>
+#include <filesystem>
+#include <optional>
+
+namespace fs = std::filesystem;
+
+#define FONTDIR "tinytex"
+
+namespace tinytex {
+
+font_paths_t GetFontPaths() {
+  std::queue<std::string> paths;
+  // checks if TINYTEX_FONTDIR exists. If it does, it pushes it to potential paths.
+  char* devdir = getenv("TINYTEX_FONTDIR");
+  if (devdir && *devdir)
+    paths.push(std::string(devdir));
+
+  // checks if XDG_DATA_HOME exists. If it does, it pushes it to potential paths.
+  char* userdata = getenv("XDG_DATA_HOME");
+  if (userdata && *userdata)
+    paths.push(std::string(userdata) + "/" + FONTDIR);
+
+  // checks if XDG_DATA_DIRS is set. If it is, it splits XDG_DATA_DIRS at : and
+  // pushes each part of it to potential paths.
+  char* data_dirs = getenv("XDG_DATA_DIRS");
+  if (data_dirs && *data_dirs) {
+    std::stringstream data_dir_paths(data_dirs);
+    std::string data_dir_path;
+    while(getline(data_dir_paths, data_dir_path, ':'))
+      paths.push(data_dir_path + "/" + FONTDIR);
+  }
+
+#ifndef _WIN32
+  char* home = getenv("HOME");
+  if (home || *home) {
+    char* userdata_fallback;
+    asprintf(&userdata_fallback, "%s/.local/share/%s/", home, FONTDIR);
+    paths.push(std::string(userdata_fallback));
+    free(userdata_fallback);
+  }
+
+  paths.push("/usr/local/share/" FONTDIR);
+  paths.push("/usr/share/" FONTDIR);
+#endif
```

Comment:
the problem is, that this will only work when the cwd is also the one in which the executable is located. what I'd prefer is that you could run the executable from anywhere (i.e. cwd is `Z:\home\florian\` but the binary is located in `C:\Program Files\Application\binary.exe`. To do that you need that function winapi has to figure out where the binary is located (similar to `dirname $(readlink("/proc/self/exe")` on POSIX) that glib uses too when getting the data dirs on windows.

Reply:
new solution

```cpp
#ifdef _WIN32
#include <windows.h>
// Get dir of executable
// For example, the full path is c:\x\y\z\a.exe
// return c:\x\y\z\
//
std::string getDirOfExecutable() {
  char szBuff[MAX_PATH] = {0};
  // WIN API: GetModuleFileNameA
  // retrieves the path of the executable file of the current process
  // see https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-getmodulefilenamea
  GetModuleFileNameA(nullptr,szBuff, MAX_PATH);
  std::string exePath(szBuff);
  auto pos = exePath.find_last_of('\\');
  // pos + 1 for adding last \ to path
  return exePath.substr(0, pos + 1);
}
#endif
```
and then

```cpp
#else
  auto exeDir = getDirOfExecutable();
  paths.push(exeDir + std::string("share/" FONTDIR));
#endif
```

it works with such file tree
```
PS C:\code\cLaTeXMath\cmake-build-debug> tree /f    
Folder PATH listing
Volume serial number is 6659-D38D
C:.
│  tinytex-qt-example.exe
│  tinytex-qt-example.ilk
│  tinytex-qt-example.pdb
│  tinytex-qt.dll
│  tinytex.dll
│  tinytex.ilk
│  tinytex.pdb
│
└─share
    └─tinytex
            XITSMath-Regular.clm
            XITSMath-Regular.otf
```
test case
```
PS C:\code\cLaTeXMath\cmake-build-debug> .\tinytex-qt-example.exe C:\code\cLaTeXMath\res\SAMPLES.tex
PS C:\code\cLaTeXMath> .\cmake-build-debug\tinytex-qt-example.exe C:\code\cLaTeXMath\res\SAMPLES.tex
PS C:\code> .\cLaTeXMath\cmake-build-debug\tinytex-qt-example.exe C:\code\cLaTeXMath\res\SAMPLES.tex
PS D:\> c:\code\cLaTeXMath\cmake-build-debug\tinytex-qt-example.exe C:\code\cLaTeXMath\res\SAMPLES.tex
```